File: stations/hook-observatory/handlers/anvil_telemetry.py
"""
Anvil telemetry — multi-channel usage tracker.

Tracks three invocation channels:
  1. Skill tool calls (PostToolUse/Skill)
  2. MCP server calls (PostToolUse/mcp__mcpproxy__call_tool_*)
  3. CLI commands (PostToolUse/Bash) — known station CLIs only

API-first, file-fallback: try synchronous POST to Anvil API first.
Only writes to local JSONL spool when API is unreachable.
SessionStart triggers background sync of any pending spool entries.

Latency: ~2-5ms when API is up (match + HTTP), <0.01ms on non-match (dict lookup).
"""

from __future__ import annotations

import json
import os
import re
import sys
import time
import urllib.error
import urllib.request

from .base import ALLOW, HookResult, run_background

ANVIL_API = os.environ.get("ANVIL_API", "http://127.0.0.1:10301")
SPOOL_DIR = os.path.join(os.path.expanduser("~"), ".claude", "data", "anvil-telemetry")
SPOOL_FILE = os.path.join(SPOOL_DIR, "pending.jsonl")

# ---------------------------------------------------------------------------
# Skill-level config
# ---------------------------------------------------------------------------

ALIAS_MAP = {
    "r": "prompt-router",
}

_TEST_PREFIXES = ("_", "test-")
_TEST_EXACT = {"test-skill", "test-verify", "general-purpose", "commit"}
_TEST_PATTERN_DIGITS = re.compile(r"^skill-\d+$")

# CLI builtins (not skills, skip intent tracking)
_CLI_BUILTINS = frozenset(
    {
        "clear",
        "exit",
        "context",
        "mcp",
        "login",
        "model",
        "config",
        "help",
        "compact",
        "fast",
        "cost",
        "memory",
        "permissions",
        "agents",
        "skills",
        "terminal-setup",
        "vim",
        "bug",
        "doctor",
        "release-notes",
        "init",
        "review",
        "allowed-tools",
        "listen",
        "status-bar",
        "add-dir",
        "loop",
    }
)

_COMMAND_NAME_RE = re.compile(r"<command-name>/([^<]+)</command-name>")

# ---------------------------------------------------------------------------
# Tool registry (loaded once at module level)
# ---------------------------------------------------------------------------

_REGISTRY_PATH = os.path.join(os.path.dirname(__file__), "tool_registry.json")

_MCP_REGISTRY: dict[str, str] = {}
_CLI_REGISTRY: dict[str, str] = {}

try:
    with open(_REGISTRY_PATH) as _f:
        _reg = json.load(_f)
    _MCP_REGISTRY = _reg.get("mcp_servers", {})
    _CLI_REGISTRY = _reg.get("cli_commands", {})
except (OSError, json.JSONDecodeError):
    pass

_MCP_PROXY_TOOLS = frozenset(
    {
        "mcp__mcpproxy__call_tool_read",
        "mcp__mcpproxy__call_tool_write",
        "mcp__mcpproxy__call_tool_destructive",
    }
)


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def _is_test(name: str) -> bool:
    """Return True if this is a test/probe invocation that should not be tracked."""
    if any(name.startswith(p) for p in _TEST_PREFIXES):
        return True
    if name in _TEST_EXACT:
        return True
    if _TEST_PATTERN_DIGITS.match(name):
        return True
    return False


def _parse_context(raw_input: str) -> dict:
    """Extract session context from raw hook input."""
    try:
        parsed = json.loads(raw_input) if raw_input.strip() else {}
    except (json.JSONDecodeError, AttributeError):
        parsed = {}
    return parsed.get("data", parsed)


def _post_intent(payload: dict) -> None:
    """POST intent to Anvil API. Fire-and-forget, no spool."""
    try:
        url = f"{ANVIL_API}/api/anvil/intents"
        if not url.startswith(("http://", "https://")):
            return
        data = json.dumps(payload, ensure_ascii=False).encode("utf-8")
        req = urllib.request.Request(  # noqa: S310
            url,
            data=data,
            headers={"Content-Type": "application/json"},
            method="POST",
        )
        urllib.request.urlopen(req, timeout=3)  # noqa: S310
    except (urllib.error.URLError, OSError, ValueError):
        pass


def _send(payload: dict) -> None:
    """POST to Anvil API, fall back to spool on failure."""
    if _post_to_api(payload):
        return
    _write_spool(payload)


# ---------------------------------------------------------------------------
# Channel handlers
# ---------------------------------------------------------------------------


def _handle_intent(raw_input: str) -> HookResult:
    """UserPromptSubmit: extract <command-name> tags and POST intents."""
    matches = _COMMAND_NAME_RE.findall(raw_input)
    if not matches:
        return ALLOW

    session_id = ""
    try:
        parsed = json.loads(raw_input) if raw_input.strip().startswith("{") else {}
        data = parsed.get("data", parsed)
        session_id = data.get("session_id", "")
    except (json.JSONDecodeError, AttributeError):
        pass

    for skill_name in matches:
        skill_name = skill_name.strip()
        if skill_name in _CLI_BUILTINS:
            continue
        if _is_test(skill_name):
            continue
        _post_intent({"skill_name": skill_name, "session_id": session_id})

    return ALLOW


def _handle_skill(tool_input: dict, raw_input: str) -> HookResult:
    """Handle Skill tool invocations."""
    skill_name = tool_input.get("skill", "")
    if not skill_name:
        return ALLOW

    if _is_test(skill_name):
        return ALLOW

    original_name = None
    if skill_name in ALIAS_MAP:
        original_name = skill_name
        skill_name = ALIAS_MAP[skill_name]

    data = _parse_context(raw_input)
    tool_response = data.get("tool_response", {})

    payload_data: dict = {
        "args": tool_input.get("args", ""),
        "cwd": data.get("cwd", ""),
    }
    if original_name:
        payload_data["original_name"] = original_name

    _send(
        {
            "skill_name": skill_name,
            "session_id": data.get("session_id", ""),
            "agent_model": data.get("agent_model", ""),
            "tool_use_id": data.get("tool_use_id", ""),
            "success": tool_response.get("success", True),
            "error_message": tool_response.get("error", None),
            "tool_calls_count": 1,
            "category": "skill",
            "payload": payload_data,
        }
    )
    return ALLOW


def _handle_mcp(tool_input: dict, raw_input: str) -> HookResult:
    """Handle MCP proxy tool calls (call_tool_read/write/destructive)."""
    # tool_input.name = "tmux-relay:relay_run"
    mcp_name = tool_input.get("name", "")
    if ":" not in mcp_name:
        return ALLOW

    server_name, tool_name = mcp_name.split(":", 1)
    station = _MCP_REGISTRY.get(server_name)
    if not station:
        return ALLOW

    data = _parse_context(raw_input)

    _send(
        {
            "skill_name": station,
            "session_id": data.get("session_id", ""),
            "agent_model": data.get("agent_model", ""),
            "tool_use_id": data.get("tool_use_id", ""),
            "success": True,
            "tool_calls_count": 1,
            "category": "mcp",
            "payload": {
                "tool": tool_name,
                "server": server_name,
                "cwd": data.get("cwd", ""),
            },
        }
    )
    return ALLOW


def _handle_cli(tool_input: dict, raw_input: str) -> HookResult:
    """Handle Bash commands that match known station CLIs."""
    cmd = tool_input.get("command", "").strip()
    if not cmd:
        return ALLOW

    # Extract first token basename: "/Users/.../relay" → "relay"
    tokens = cmd.split(None, 2)  # split into max 3 parts
    if not tokens:
        return ALLOW

    binary = os.path.basename(tokens[0])
    station = _CLI_REGISTRY.get(binary)
    if not station:
        return ALLOW

    data = _parse_context(raw_input)
    subcommand = tokens[1] if len(tokens) > 1 else ""

    _send(
        {
            "skill_name": station,
            "session_id": data.get("session_id", ""),
            "agent_model": data.get("agent_model", ""),
            "tool_use_id": data.get("tool_use_id", ""),
            "success": True,
            "tool_calls_count": 1,
            "category": "cli",
            "payload": {
                "tool": subcommand,
                "command": cmd[:200],
                "cwd": data.get("cwd", ""),
            },
        }
    )
    return ALLOW


# ---------------------------------------------------------------------------
# Main entry point
# ---------------------------------------------------------------------------


def handle(event_type: str, tool_name: str, tool_input: dict, raw_input: str) -> HookResult:
    """Multi-channel telemetry handler.

    Routes to the appropriate channel handler based on tool_name.
    Non-matching tools return ALLOW immediately (<0.01ms).
    """
    if event_type == "SessionStart":
        return _sync_pending()

    if event_type == "UserPromptSubmit":
        return _handle_intent(raw_input)

    if tool_name == "Skill":
        return _handle_skill(tool_input, raw_input)

    if tool_name in _MCP_PROXY_TOOLS:
        return _handle_mcp(tool_input, raw_input)

    if tool_name == "Bash":
        return _handle_cli(tool_input, raw_input)

    return ALLOW


# ---------------------------------------------------------------------------
# API + spool (shared infrastructure)
# ---------------------------------------------------------------------------


def _post_to_api(payload: dict) -> bool:
    """Synchronous POST to Anvil API. Returns True on success."""
    try:
        url = f"{ANVIL_API}/api/anvil/invocations"
        if not url.startswith(("http://", "https://")):
            return False
        data = json.dumps(payload, ensure_ascii=False).encode("utf-8")
        req = urllib.request.Request(  # noqa: S310
            url,
            data=data,
            headers={"Content-Type": "application/json"},
            method="POST",
        )
        with urllib.request.urlopen(req, timeout=3) as resp:  # noqa: S310
            return resp.status in (200, 201)
    except (urllib.error.URLError, OSError, ValueError) as exc:
        print(f"[anvil-telemetry] API POST failed: {exc}", file=sys.stderr)
        return False


def _write_spool(payload: dict) -> None:
    """Append invocation to local JSONL spool. Never raises."""
    try:
        os.makedirs(SPOOL_DIR, exist_ok=True)
        entry = {
            "ts": time.strftime("%Y-%m-%dT%H:%M:%S", time.localtime()),
            "synced": False,
            **payload,
        }
        with open(SPOOL_FILE, "a") as f:
            f.write(json.dumps(entry, ensure_ascii=False) + "\n")
    except OSError as exc:
        print(f"[anvil-telemetry] Spool write failed: {exc}", file=sys.stderr)


def _sync_pending() -> HookResult:
    """SessionStart: background sync any pending spool entries."""
    if not os.path.exists(SPOOL_FILE):
        return ALLOW
    try:
        with open(SPOOL_FILE) as f:
            has_pending = any(line.strip() for line in f)
        if has_pending:
            run_background(
                [
                    os.path.join(os.path.expanduser("~"), ".local", "bin", "python3"),
                    os.path.join(
                        os.path.expanduser("~"),
                        "workshop",
                        "stations",
                        "anvil",
                        "scripts",
                        "anvil_telemetry_sync.py",
                    ),
                ]
            )
    except OSError:
        pass
    return ALLOW

File: internal/handlers/anvil_telemetry.go
package handlers

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"time"

	"github.com/joneshong/hook-observatory/internal/core"
)

func init() {
	entry := core.Entry{
		Handler:    anvilTelemetryHandle,
		ModuleName: "anvil_telemetry",
	}
	core.Register("UserPromptSubmit", entry)
	core.Register("PostToolUse", entry)
	core.Register("SessionStart", entry)
}

// ---------------------------------------------------------------------------
// Constants & registry
// ---------------------------------------------------------------------------

var (
	anvilAliasMap = map[string]string{
		"r": "prompt-router",
	}
	anvilTestPrefixes = []string{"_", "test-"}
	anvilTestExact    = map[string]bool{
		"test-skill": true, "test-verify": true,
		"general-purpose": true, "commit": true,
	}
	anvilTestDigitsRe  = regexp.MustCompile(`^skill-\d+$`)
	anvilCommandNameRe = regexp.MustCompile(`<command-name>/([^<]+)</command-name>`)
	anvilMCPProxyTools = map[string]bool{
		"mcp__mcpproxy__call_tool_read":        true,
		"mcp__mcpproxy__call_tool_write":       true,
		"mcp__mcpproxy__call_tool_destructive": true,
	}
	anvilCLIBuiltins = map[string]bool{
		"clear": true, "exit": true, "context": true, "mcp": true,
		"login": true, "model": true, "config": true, "help": true,
		"compact": true, "fast": true, "cost": true, "memory": true,
		"permissions": true, "agents": true, "skills": true,
		"terminal-setup": true, "vim": true, "bug": true, "doctor": true,
		"release-notes": true, "init": true, "review": true,
		"allowed-tools": true, "listen": true, "status-bar": true,
		"add-dir": true, "loop": true,
	}

	anvilMCPRegistry map[string]string
	anvilCLIRegistry map[string]string
)

func init() {
	// Load tool_registry.json from the hook-observatory assets dir.
	// HOOK_OBSERVATORY_ROOT (preferred) / HOOK_DISPATCHER_ROOT (backward-compat) override default.
	root := os.Getenv("HOOK_OBSERVATORY_ROOT")
	if root == "" {
		root = os.Getenv("HOOK_DISPATCHER_ROOT")
	}
	if root == "" {
		home, _ := os.UserHomeDir()
		root = filepath.Join(home, "workshop", "stations", "hook-observatory")
	}
	regPath := filepath.Join(root, "assets", "tool_registry.json")
	if data, err := os.ReadFile(regPath); err == nil {
		var reg struct {
			MCPServers  map[string]string `json:"mcp_servers"`
			CLICommands map[string]string `json:"cli_commands"`
		}
		if json.Unmarshal(data, &reg) == nil {
			anvilMCPRegistry = reg.MCPServers
			anvilCLIRegistry = reg.CLICommands
		}
	}
	if anvilMCPRegistry == nil {
		anvilMCPRegistry = map[string]string{}
	}
	if anvilCLIRegistry == nil {
		anvilCLIRegistry = map[string]string{}
	}
}

func anvilSpoolDir() string {
	dataDir := core.Cfg().GetPath("data_dir")
	if dataDir != "" {
		return filepath.Join(dataDir, "anvil-telemetry")
	}
	home, _ := os.UserHomeDir()
	return filepath.Join(home, ".claude", "data", "anvil-telemetry")
}

func anvilSpoolFile() string {
	return filepath.Join(anvilSpoolDir(), "pending.jsonl")
}

func anvilAPIURL() string {
	url := os.Getenv("ANVIL_API")
	if url == "" {
		url = core.Cfg().GetService("anvil_url")
	}
	return url
}

// ---------------------------------------------------------------------------
// Main handler
// ---------------------------------------------------------------------------

func anvilTelemetryHandle(eventType, toolName string, toolInput map[string]any, rawInput string) core.HookResult {
	switch eventType {
	case "SessionStart":
		return anvilSyncPending()
	case "UserPromptSubmit":
		return anvilHandleIntent(rawInput)
	}
	// PostToolUse
	if toolName == "Skill" {
		return anvilHandleSkill(toolInput, rawInput)
	}
	if anvilMCPProxyTools[toolName] {
		return anvilHandleMCP(toolInput, rawInput)
	}
	if toolName == "Bash" {
		return anvilHandleCLI(toolInput, rawInput)
	}
	return core.Allow()
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

func anvilIsTest(name string) bool {
	for _, p := range anvilTestPrefixes {
		if strings.HasPrefix(name, p) {
			return true
		}
	}
	if anvilTestExact[name] {
		return true
	}
	return anvilTestDigitsRe.MatchString(name)
}

func anvilParseContext(rawInput string) map[string]any {
	if strings.TrimSpace(rawInput) == "" {
		return map[string]any{}
	}
	var parsed map[string]any
	if err := json.Unmarshal([]byte(rawInput), &parsed); err != nil {
		return map[string]any{}
	}
	if data, ok := parsed["data"].(map[string]any); ok {
		return data
	}
	return parsed
}

func anvilPostToAPI(endpoint string, payload map[string]any) bool {
	base := anvilAPIURL()
	if base == "" {
		return false
	}
	url := base + endpoint
	if !strings.HasPrefix(url, "http://") && !strings.HasPrefix(url, "https://") {
		return false
	}
	data, err := json.Marshal(payload)
	if err != nil {
		return false
	}
	client := &http.Client{Timeout: 3 * time.Second}
	resp, err := client.Post(url, "application/json", strings.NewReader(string(data)))
	if err != nil {
		return false
	}
	defer resp.Body.Close()
	io.Copy(io.Discard, resp.Body)
	return resp.StatusCode == 200 || resp.StatusCode == 201
}

func anvilWriteSpool(payload map[string]any) {
	dir := anvilSpoolDir()
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return
	}
	entry := map[string]any{
		"ts":     time.Now().Format("2006-01-02T15:04:05"),
		"synced": false,
	}
	for k, v := range payload {
		entry[k] = v
	}
	b, err := json.Marshal(entry)
	if err != nil {
		return
	}
	f, err := os.OpenFile(anvilSpoolFile(), os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o644)
	if err != nil {
		return
	}
	defer f.Close()
	fmt.Fprintf(f, "%s\n", b)
}

func anvilSend(payload map[string]any) {
	if anvilPostToAPI("/api/anvil/invocations", payload) {
		return
	}
	anvilWriteSpool(payload)
}

// ---------------------------------------------------------------------------
// Channel handlers
// ---------------------------------------------------------------------------

func anvilHandleIntent(rawInput string) core.HookResult {
	matches := anvilCommandNameRe.FindAllStringSubmatch(rawInput, -1)
	if len(matches) == 0 {
		return core.Allow()
	}
	sessionID := ""
	var parsed map[string]any
	if json.Unmarshal([]byte(rawInput), &parsed) == nil {
		data := parsed
		if d, ok := parsed["data"].(map[string]any); ok {
			data = d
		}
		if s, ok := data["session_id"].(string); ok {
			sessionID = s
		}
	}
	for _, m := range matches {
		name := strings.TrimSpace(m[1])
		if anvilCLIBuiltins[name] || anvilIsTest(name) {
			continue
		}
		anvilPostToAPI("/api/anvil/intents", map[string]any{
			"skill_name": name,
			"session_id": sessionID,
		})
	}
	return core.Allow()
}

func anvilHandleSkill(toolInput map[string]any, rawInput string) core.HookResult {
	skillName, _ := toolInput["skill"].(string)
	if skillName == "" || anvilIsTest(skillName) {
		return core.Allow()
	}
	originalName := ""
	if alias, ok := anvilAliasMap[skillName]; ok {
		originalName = skillName
		skillName = alias
	}
	data := anvilParseContext(rawInput)
	toolResponse, _ := data["tool_response"].(map[string]any)
	success := true
	var errMsg interface{}
	if toolResponse != nil {
		if v, ok := toolResponse["success"].(bool); ok {
			success = v
		}
		if e, ok := toolResponse["error"]; ok {
			errMsg = e
		}
	}
	payloadData := map[string]any{
		"args": toolInput["args"],
		"cwd":  data["cwd"],
	}
	if originalName != "" {
		payloadData["original_name"] = originalName
	}
	anvilSend(map[string]any{
		"skill_name":       skillName,
		"session_id":       data["session_id"],
		"agent_model":      data["agent_model"],
		"tool_use_id":      data["tool_use_id"],
		"success":          success,
		"error_message":    errMsg,
		"tool_calls_count": 1,
		"category":         "skill",
		"payload":          payloadData,
	})
	return core.Allow()
}

func anvilHandleMCP(toolInput map[string]any, rawInput string) core.HookResult {
	mcpName, _ := toolInput["name"].(string)
	idx := strings.Index(mcpName, ":")
	if idx < 0 {
		return core.Allow()
	}
	serverName := mcpName[:idx]
	toolName := mcpName[idx+1:]
	station, ok := anvilMCPRegistry[serverName]
	if !ok {
		return core.Allow()
	}
	data := anvilParseContext(rawInput)
	anvilSend(map[string]any{
		"skill_name":       station,
		"session_id":       data["session_id"],
		"agent_model":      data["agent_model"],
		"tool_use_id":      data["tool_use_id"],
		"success":          true,
		"tool_calls_count": 1,
		"category":         "mcp",
		"payload": map[string]any{
			"tool":   toolName,
			"server": serverName,
			"cwd":    data["cwd"],
		},
	})
	return core.Allow()
}

func anvilHandleCLI(toolInput map[string]any, rawInput string) core.HookResult {
	cmd, _ := toolInput["command"].(string)
	cmd = strings.TrimSpace(cmd)
	if cmd == "" {
		return core.Allow()
	}
	tokens := strings.Fields(cmd)
	if len(tokens) == 0 {
		return core.Allow()
	}
	binary := filepath.Base(tokens[0])
	station, ok := anvilCLIRegistry[binary]
	if !ok {
		return core.Allow()
	}
	data := anvilParseContext(rawInput)
	subcommand := ""
	if len(tokens) > 1 {
		subcommand = tokens[1]
	}
	cmdTrunc := cmd
	if len(cmdTrunc) > 200 {
		cmdTrunc = cmdTrunc[:200]
	}
	anvilSend(map[string]any{
		"skill_name":       station,
		"session_id":       data["session_id"],
		"agent_model":      data["agent_model"],
		"tool_use_id":      data["tool_use_id"],
		"success":          true,
		"tool_calls_count": 1,
		"category":         "cli",
		"payload": map[string]any{
			"tool":    subcommand,
			"command": cmdTrunc,
			"cwd":     data["cwd"],
		},
	})
	return core.Allow()
}

func anvilSyncPending() core.HookResult {
	spoolFile := anvilSpoolFile()
	if _, err := os.Stat(spoolFile); os.IsNotExist(err) {
		return core.Allow()
	}
	// Run sync in background goroutine — same pattern as externalSkillTracker
	go anvilSyncPendingInProc()
	return core.Allow()
}

// anvilSyncPendingInProc is the Go in-process port of anvil_telemetry_sync.py.
// Reads pending.jsonl, POSTs unsent entries to Anvil API, moves synced entries
// to synced.jsonl, and rewrites pending.jsonl with only failed entries.
func anvilSyncPendingInProc() {
	spoolFile := anvilSpoolFile()
	raw, err := os.ReadFile(spoolFile)
	if err != nil {
		return
	}

	type spoolEntry = map[string]any

	var allEntries []spoolEntry
	for _, line := range strings.Split(string(raw), "\n") {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}
		var entry spoolEntry
		if json.Unmarshal([]byte(line), &entry) == nil {
			allEntries = append(allEntries, entry)
		}
	}

	if len(allEntries) == 0 {
		return
	}

	base := anvilAPIURL()
	dir := anvilSpoolDir()
	syncedFile := filepath.Join(dir, "synced.jsonl")

	var stillPending []spoolEntry
	var newlySynced []spoolEntry

	client := &http.Client{Timeout: 5 * time.Second}

	for _, entry := range allEntries {
		if synced, _ := entry["synced"].(bool); synced {
			// Already marked synced — move to archive
			newlySynced = append(newlySynced, entry)
			continue
		}

		sent := false
		if base != "" && (strings.HasPrefix(base, "http://") || strings.HasPrefix(base, "https://")) {
			// Build payload matching Python's post_to_anvil()
			payload := map[string]any{
				"skill_name":       entry["skill_name"],
				"session_id":       entry["session_id"],
				"agent_model":      entry["agent_model"],
				"success":          entry["success"],
				"error_message":    entry["error_message"],
				"tool_calls_count": entry["tool_calls_count"],
				"payload":          entry["payload"],
			}
			data, err := json.Marshal(payload)
			if err == nil {
				resp, err := client.Post(base+"/api/anvil/invocations", "application/json", strings.NewReader(string(data)))
				if err == nil {
					io.Copy(io.Discard, resp.Body)
					resp.Body.Close()
					// Match Python _post_to_api: accept both 200 and 201 as success.
					// Otherwise a 200 response would leave entries spooled forever.
					if resp.StatusCode == 200 || resp.StatusCode == 201 {
						sent = true
					}
				}
			}
		}

		if sent {
			entry["synced"] = true
			newlySynced = append(newlySynced, entry)
		} else {
			stillPending = append(stillPending, entry)
		}
	}

	// Append synced entries to archive file
	if len(newlySynced) > 0 {
		if err := os.MkdirAll(dir, 0o755); err == nil {
			sf, err := os.OpenFile(syncedFile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o644)
			if err == nil {
				for _, e := range newlySynced {
					b, _ := json.Marshal(e)
					fmt.Fprintf(sf, "%s\n", b)
				}
				sf.Close()
			}
		}
	}

	// Rewrite pending.jsonl with only unsynced entries
	var pendingLines []string
	for _, e := range stillPending {
		b, _ := json.Marshal(e)
		pendingLines = append(pendingLines, string(b))
	}
	content := ""
	if len(pendingLines) > 0 {
		content = strings.Join(pendingLines, "\n") + "\n"
	}
	_ = os.WriteFile(spoolFile, []byte(content), 0o644)
}
